File: AI_Model/AI_Model2.js
const { google } = require('googleapis');
const { GoogleAuth } = require('google-auth-library');
const { PredictionServiceClient } = require('@google-cloud/aiplatform').v1;
const { ImageClassificationPredictionInstance, ImageClassificationPredictionParams } = require('@google-cloud/aiplatform').protos.google.cloud.aiplatform.v1.schema.predict;
const fs = require('fs');
const path = require('path');
const base64 = require('base64-js');

// Initialize the Google Auth library
const keyPath = path.join(__dirname, '../AI_Model/lunar-parsec-417813-9b4958f8545d.json');
const auth = new GoogleAuth({
    keyFile: keyPath,
    scopes: ['https://www.googleapis.com/auth/cloud-platform'],
});

async function main() {
    const credentials = await auth.getClient();

    if (credentials.expired) {
        await credentials.refresh();
    }

    const PROJECT_ID = 'lunar-parsec-417813';
    const REGION = 'us-central1';

    process.env.GOOGLE_APPLICATION_CREDENTIALS = keyPath;

    const aiplatform = new PredictionServiceClient({
        projectId: PROJECT_ID,
        keyFilename: keyPath,
    });

    async function predictImageClassificationSample(project, endpointId, filename, location = 'us-central1', apiEndpoint = 'us-central1-aiplatform.googleapis.com') {
        // The AI Platform services require regional API endpoints.
        const clientOptions = { apiEndpoint: apiEndpoint };

        // Initialize client that will be used to create and send requests.
        const client = new PredictionServiceClient(clientOptions);

        const fileContent = fs.readFileSync(filename);
        const encodedContent = base64.fromByteArray(new Uint8Array(fileContent)).toString('utf-8');

        const instance = ImageClassificationPredictionInstance.create({
            content: encodedContent,
        });

        const instances = [instance];
        const parameters = ImageClassificationPredictionParams.create({
            confidenceThreshold: 0.5,
            maxPredictions: 5,
        });

        const endpoint = client.endpointPath({
            project: project,
            location: location,
            endpoint: endpointId,
        });

        const request = {
            endpoint: endpoint,
            instances: instances,
            parameters: parameters,
        };

        const [response] = await client.predict(request);

        console.log('response');
        console.log('deployed_model_id:', response.deployedModelId);

        const predictions = response.predictions;
        for (const prediction of predictions) {
            console.log('prediction:', prediction.toObject());
        }
    }

    // Predict
    predictImageClassificationSample(
        '372352180346',
        '962774162722521088',
        path.join(__dirname, 'descarga-89-_jpeg_jpg.rf.885415b610869fdae91c2ce142ace0b6.jpg')
    );
}

main().catch(console.error);

File: AI_Model/AI_Model2.py
import base64
import requests
from google.cloud import aiplatform
from google.oauth2.service_account import Credentials
from google.auth.transport.requests import Request
import os

# Get the directory of the current script
current_dir = os.path.dirname(__file__)

# Construct the path to the key file relative to the current script
key_path = os.path.join(current_dir, '../AI_Model/lunar-parsec-417813-9b4958f8545d.json')

# Normalize the path
key_path = os.path.normpath(key_path)


def predict_from_url(
    image_url: str,
    key_path=key_path,
    project: str = "372352180346",
    endpoint_id: str = "962774162722521088",
    location: str = "us-central1",
    api_endpoint: str = "us-central1-aiplatform.googleapis.com",
):

    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = key_path

    credentials = Credentials.from_service_account_file(key_path)

    if credentials.expired:
        credentials.refresh(Request())

    aiplatform.init(
        project=project,
        location=location,
        credentials=credentials
    )

    client_options = {"api_endpoint": api_endpoint}
    client = aiplatform.gapic.PredictionServiceClient(client_options=client_options)

    # Fetch image content from the URL
    response = requests.get(image_url)
    if response.status_code != 200:
        raise ValueError("Failed to fetch image from the provided URL.")

    file_content = response.content

    encoded_content = base64.b64encode(file_content).decode("utf-8")
    instance = aiplatform.gapic.schema.predict.instance.ImageClassificationPredictionInstance(
        content=encoded_content,
    ).to_value()
    instances = [instance]

    parameters = aiplatform.gapic.schema.predict.params.ImageClassificationPredictionParams(
        confidence_threshold=0.4,
        max_predictions=5,
    ).to_value()

    endpoint = client.endpoint_path(
        project=project, location=location, endpoint=endpoint_id
    )
    response = client.predict(
        endpoint=endpoint, instances=instances, parameters=parameters
    )

    print("response")
    print(" deployed_model_id:", response.deployed_model_id)

    predictions = response.predictions
    for prediction in predictions:
        print(" prediction:", dict(prediction))

# Make a prediction using the predict_from_url function with a sample image URL
predict_from_url(
    image_url="https://tse3.mm.bing.net/th?id=OIP.A82KyPwxm--kQzdzDVVnIwHaEK&pid=Api&P=0&h=180"
)
